lib/spack/llnl/util/filesystem.py

```diff
@@ -817,7 +823,22 @@ def escaped_path(path):
                             tty.debug("Redirecting link {0} to {1}".format(target, new_target))
                             target = new_target
 
-                    symlink(target, d)
+                    ignore = ignore or (lambda filename: False)
+
+                    # Build an accumulating list of all parent file paths to check against the
+                    # ignores func to make sure the source file is going to be copied into the new
+                    # tree before adding the symlink to the list of links to make after files
+                    # have been copied.
+                    all_parents = accumulate(target.split(os.sep), lambda x, y: os.path.join(x, y))
+                    if any(map(ignore, all_parents)):
+                        tty.msg(
+                            f"Ignoring link {d} because the source or a part of the source's "
+                            f"path is included in the ignores."
+                        )
```

Comment:
This should be an update to `traverse_tree`: add a `follow_ignore` parameter (which defaults to `True` since it seems that's what people have been using everywhere else).

Reply:
Giving the user the option to turn `follow_ignore` off  would cause a `SymlinkError` to be raised because the resulting link would be broken.